### PR TITLE
An example using hdf5 references (#396)

### PR DESCRIPTION
**Pull request by plusangel**
I included an example to demonstrate how references work in HighFive. PR suggested by @tdegeus 

Thank you guys

### CODE REVIEW

**Review by tdegeus — COMMENTED** [review]
I'm probably not the best person to review, because I don't actually know references in HDF5. That being said, it didn't become clear from this example either... Some line of text would surely help there: describe what the example is doing in a few lines of comment somewhere. 

Where I really got lost though is with the `  std::vector<Reference>`. The question is: Are reference in HDF5 something similar to those in C++? If so, wouldn't it be easier to (also) have a basic example with one reference?

File: src/examples/read_write_vector_dataset_references.cpp [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+#include <iostream>
+#include <string>
+#include <vector>
+
+#include <highfive/H5DataSet.hpp>
+#include <highfive/H5DataSpace.hpp>
+#include <highfive/H5File.hpp>
+#include <highfive/H5Reference.hpp>
+
+using namespace HighFive;
```
[the remark this answers is not in the document]
Comment (ohm314) [reply src/examples/read_write_vector_dataset_references.cpp:18]: I agree, please remove the using namespace declaration. As @tdegeus said, this just makes things less clear and is generally a bad practice that should be avoided.

File: src/examples/read_write_vector_dataset_references.cpp [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+const std::string FILE_NAME("dataset_integer.h5");
+const std::string SOURCE_INT_DATASET_NAME("source_dateset");
```
Comment (ohm314) [inline src/examples/read_write_vector_dataset_references.cpp:21]:
```suggestion
const std::string SOURCE_INT_DATASET_NAME("source_dataset");
```

File: src/examples/read_write_vector_dataset_references.cpp [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+const std::string REFERENCE_DATASET_NAME("reference_dateset");
```
[the remark this answers is not in the document]
Comment (ohm314) [reply src/examples/read_write_vector_dataset_references.cpp:22]: I don't mind either. Defining the strings doesn't hurt, but doesn't add much. Either is fine.

File: src/examples/read_write_vector_dataset_references.cpp [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+const size_t size_dataset = 20;
+
+// create a dataset 1D from a vector of string
```
Comment (ohm314) [inline src/examples/read_write_vector_dataset_references.cpp:25]:
```suggestion
// create a dataset 1D from a vector of int
```

File: src/examples/read_write_vector_dataset_references.cpp [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+void write_dataset() {
+
+    // we create a new hdf5 file
+    File file(FILE_NAME, File::ReadWrite | File::Create | File::Truncate);
+
+    // we create a new group
+    Group group = file.createGroup("a_group");
```
[the remark this answers is not in the document]
Comment (ohm314) [reply src/examples/read_write_vector_dataset_references.cpp:32]: why do you define const strings for the file name and dataset names, but not the group. I would be consistent, whatever you choose.

File: src/examples/read_write_vector_dataset_references.cpp [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    // let's create a dataset of native integer with the size of the vector
+    // 'data' inside the group
+    DataSet dataset = group.createDataSet<int>(SOURCE_INT_DATASET_NAME, DataSpace::From(data));
+    dataset.write(data);
+
+    // create a reference to the dateset containing the integers
+    Reference ref = Reference(group, dataset);
+    std::vector<Reference> ref_container{ref};
+
+    // in similar fashion, we store as dataset the reference that we want
```
Comment (ohm314) [inline src/examples/read_write_vector_dataset_references.cpp:48]:
```suggestion
    // in similar fashion, we store as dataset the vector of references that we want
```

**Review by ohm314 — CHANGES_REQUESTED** [review]
Hi @plusangel, sorry for taking so long to review this. Overall the example is small enough and a good idea to add. We actually have in the unit tests already something very similar (but I understand that looking at test cases is a pain, especially when you're just getting into the library).
@tdegeus gave already some very good feedback. Could I ask you to please implement his and my suggestions into an additional patch. I"ll try to then re-review fairly quickly and have this merged!
Here's the test case, for reference: https://github.com/BlueBrain/HighFive/blob/master/tests/unit/tests_high_five_base.cpp#L1587

**Comment by plusangel** [comment]
Thank you both for your valuable comments and your time to review it. I think I catch them all. Keep up the good work. Thank you!

File: src/examples/read_write_vector_dataset_references.cpp [hunk 7]
```diff
@@ -0,0 +1,97 @@
+/*
+ *  Copyright (c), 2017, Adrien Devresse
```
Comment (ferdonline) [inline src/examples/read_write_vector_dataset_references.cpp:2]:
```suggestion
 *  Copyright (c), 2021 Blue Brain Project - EPFL
```

File: src/examples/read_write_vector_dataset_references.cpp [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+#include <highfive/H5DataSet.hpp>
+#include <highfive/H5DataSpace.hpp>
+#include <highfive/H5File.hpp>
+#include <highfive/H5Reference.hpp>
+
+// const std::string FILE_NAME("dataset_integer.h5");
+//const std::string SOURCE_INT_DATASET_NAME("source_dateset");
+//const std::string REFERENCE_DATASET_NAME("reference_dataset");
+//const size_t size_dataset = 20;
+
```
Comment (ferdonline) [inline src/examples/read_write_vector_dataset_references.cpp:22]:
```suggestion
```

File: src/examples/read_write_vector_dataset_references.cpp [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+// create a dataset 1D from a vector of int
+void write_dataset() {
+
+    // we create a new hdf5 file
+    HighFive::File file("dataset_integer.h5", HighFive::File::ReadWrite |
+                                                  HighFive::File::Create |
+                                                  HighFive::File::Truncate);
```
Comment (ferdonline) [inline src/examples/read_write_vector_dataset_references.cpp:29]:
```suggestion
    HighFive::File file("dataset_integer.h5", HighFive::File::Overwrite);
```

File: src/examples/read_write_vector_dataset_references.cpp [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+
+    // we create a new group
+    HighFive::Group group = file.createGroup("a_group");
+
+    std::vector<int> data(20);
+    for (size_t i = 0; i < data.size(); ++i) {
+        data[i] = int(i);
+    }
```
Comment (ferdonline) [inline src/examples/read_write_vector_dataset_references.cpp:37]:
```suggestion
    std::vector<int> data(20);
    std::iota(data.begin(), data.end(), 0);
```

File: src/examples/read_write_vector_dataset_references.cpp [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+
+    // let's create a dataset of native integer with the size of the vector
+    // 'data' inside the group
+    HighFive::DataSet dataset = group.createDataSet<int>("source_dateset",
+                                                         HighFive::DataSpace::From(data));
+    dataset.write(data);
```
Comment (ferdonline) [inline src/examples/read_write_vector_dataset_references.cpp:43]:
```suggestion
    auto dataset = group.createDataSet("source_dataset", data);
```

File: src/examples/read_write_vector_dataset_references.cpp [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+        write_dataset();
+        read_dataset();
+
+    } catch (HighFive::Exception& err) {
+        // catch and print any HDF5 error
+        std::cerr << err.what() << std::endl;
+    }
+
+    return 0;  // successfully terminated
+}
```
Comment (ferdonline) [inline src/examples/read_write_vector_dataset_references.cpp:97]:
```suggestion
}

```

**Review by ferdonline — CHANGES_REQUESTED** [review]

**Comment by ferdonline** [comment]
@plusangel Do you have an estimate on when you could address this PR suggestions?

**Comment by plusangel** [comment]
Sorry for the delay @ferdonline. Done :)

**Review by ferdonline — DISMISSED** [review]

**Review by tdegeus — APPROVED** [review]

**Review by ohm314 — APPROVED** [review]
lgtm

**Pull request merged**